tests/Config/Plugin/Hook/ArrayMapStorageProvider.php

```diff
@@ -0,0 +1,190 @@
+<?php
+
+namespace Psalm\Tests\Config\Plugin\Hook;
+
+use Psalm\Codebase;
+use Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer;
+use Psalm\Internal\Type\TemplateInferredTypeReplacer;
+use Psalm\Internal\Type\TemplateResult;
+use Psalm\Internal\Type\TemplateStandinTypeReplacer;
+use Psalm\Plugin\EventHandler\Event\FunctionDynamicStorageProviderEvent;
+use Psalm\Plugin\EventHandler\FunctionDynamicStorageProviderInterface;
+use Psalm\Storage\FunctionLikeParameter;
+use Psalm\Storage\FunctionStorage;
+use Psalm\Type;
+use Psalm\Type\Atomic\TCallable;
+use Psalm\Type\Atomic\TTemplateParam;
+use Psalm\Type\Union;
+
+use function array_keys;
+use function array_map;
+use function count;
+
+class ArrayMapStorageProvider implements FunctionDynamicStorageProviderInterface
+{
+    public static function getFunctionIds(): array
+    {
+        return ['custom_array_map'];
+    }
+
+    public static function getFunctionStorage(FunctionDynamicStorageProviderEvent $event): ?FunctionStorage
+    {
+        $context = $event->getContext();
+        $statements_analyzer = $event->getStatementsAnalyzer();
+        $call_args = $event->getCallArgs();
+        $args_count = count($call_args);
+        $expected_callable_args_count = $args_count - 1;
+
+        if ($expected_callable_args_count < 1) {
+            return null;
+        }
+
+        $last_array_arg = $call_args[$args_count - 1];
+
+        if (ExpressionAnalyzer::analyze($statements_analyzer, $last_array_arg->value, $context) === false) {
```

Comment:
Are the internal classes supposed to be used inside 3rd party plugins?
Because this would be the only way to fetch the datatype on the next line after the if statement.

Reply:
In the plugins that I wrote before, I was not bothered by this. Of course, it may be that the internal api will change, but I have never had this happen. Maybe we can think about exposing some internal api to pulbic api, but I don't know how to approach it.